Cleanup all.sh

mpg opened the pull request:
## Description
General cleanup of all.sh: fix misc small things, organise the tests in a more readable way.
Internal ref: IOTSSL-1990

This tries to achieve two things:
- various fixes and low-hanging optimisations in all.sh
- reorganise it to we get a clearer idea of what kind of tests we run, to prepare future work on rationalising our testing plan

It is likely that, as part of reworking the CI, all.sh will undergo profound transformations (ideally it would end up just invoking a particular test campaign), so the goal of the PR is not to make it perfect, so some unsatisfying things (eg section 5 at the end) are left untouched for now, as they are expected to be addressed later.

## Status
**READY**

## Requires Backporting
Yes, as it touches the test infrastructure.

## Migrations
NO

## Todo
- [x] 2.7 backport: https://github.com/ARMmbed/mbedtls/pull/1575
- [x] 2.1 backport: https://github.com/ARMmbed/mbedtls/pull/1579

gilles-peskine-arm commented:
I unknowingly redid some of this in #1390. If that isn't merged by the time you get around to rebasing/merging to 2.7 and development, I suggest that you rebase/merge on top of #1390.

mpg commented:
Note: `all.sh` passing locally as of 4d5c017.

gilles-peskine-arm commented:
The cmac fix had already been submitted as #726.

mpg commented:
Thanks for the info! At the end of this sprint, I'll rebase again on then-current development, and double-check that my cmac-fixing commit is removed.

mpg commented:
@mazimkhan @Patater I'd like to get this merged relatively soon as I'm getting tired of rebasing it (git is not very helpful when some commits move lots of code around) and I think it's good preparation for CI work which I hope/expect to be a priority soon. If you're unavailable to review this, can you please let me know so that I can ask other people? Thanks!

Patater commented on tests/scripts/all.sh:
```diff
@@ -531,8 +531,7 @@ cleanup
 scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE # too slow for tests
-CC=clang cmake -D CMAKE_BUILD_TYPE:String=Check -D ENABLE_TESTING=On .
-make
+make CC='clang' CFLAGS='-Werror -O2'
```
Was it intentional to change from cmake to make here?

mazimkhan replied:
I guess: Is it for ease to specify ```CFLAGS```.

mpg replied:
@Patater It was not intentional to do it in this commit (otherwise it would have been motivated in the commit message - probably messed up something while trying to clean up history after rebasing), but it was intentional. The motivation is that the first build uses cmake, so by using make here we test both build systems in the first two builds (just like we do first build with gcc and this one with clang). This is not a very strong motivation, but since the choice of `cmake` vs `make` is otherwise quite arbitrary here, why not?

@gilles-peskine-arm I'll fix the message.

@mazimkhan That was not really my motivation here, as `Check` is supposed to set exactly the flags we want anyway.

mpg commented:
`all.sh -k -r` passed locally with 841af73479fc

Patater commented on tests/scripts/all.sh:
```diff
@@ -467,15 +467,15 @@ tests/scripts/doxygen.sh
 #################################
 
-msg "build: cmake, gcc, ASan" # ~ 1 min 50s
+msg "build: cmake, gcc, ASan" # ~ 1 min
```
Where do these numbers come from? How are they measured? Is their benefit worth the maintenance effort?

mpg replied:
Those numbers come from the output of a successful run (`msg` prints current time) plus some math made by hand. (In theory, they're supposed to be added when adding a new test, but in practice that doesn't quite happen.)

Regarding benefit vs effort, now that you're asking, probably not. In the old times, `all.sh` used to be ordered by decreasing "perceived likelihood of finding bugs I missed during development" / "running time", but that stopped making sense a while ago. With the idea that we're probably going to rationalize our test campaigns soon, I thought it made sense to update them one last time in case the timings might guide use in that work, but in the long run I don't think those numbers are worth the effort (but if we think they are, we should make their computation more automated).

Patater commented on tests/scripts/all.sh:
```diff
@@ -150,7 +150,7 @@ cleanup()
     command make clean
 
-    find . -name yotta -prune -o -iname '*cmake*' -not -name CMakeLists.txt -exec rm -rf {} \+
+    find . -name yotta -prune -o -name .git -prune -o -iname '*cmake*' -not -name CMakeLists.txt -exec rm -rf {} \+
```
Did this actually screw anything up ever? Just curious, since it sounds like a PITA to debug this.

mpg replied:
Ask @gilles-peskine-arm ;) (I also noticed that some of my local branches kept disappearing, but never made the connection.)

gilles-peskine-arm replied:
@Patater Imagine having a git branch whose name contains `cmake`. Took me a while to understand why my head got detached. Then I [tightened this up](https://github.com/ARMmbed/mbedtls/pull/1470/commits/31b07e283321b1470c31466d10ca8a8099b73bba).

gilles-peskine-arm commented:
@mpg Did you test some failing cases?

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -810,7 +810,7 @@ make CC=gcc CFLAGS='-Werror -Os'
 
 if uname -a | grep -F Linux >/dev/null; then
-    msg "build/test: make shared" # ~ 40s
+    msg "build+test: make shared" # ~ 40s
```
This is the only line changed by a commit whose message is “fix incorrect use of if_build_succeeded”. What gives?

mpg replied:
After rebasing, I reviewed each commit and either deleted or amended the ones that no longer made sense. Apparently I should have done a second pass and checked that the commit actually made sense after amending. Sorry for that.

Should I rewrite history to drop this commit, or is it too late now that the review process has started? Or maybe I could do it after review of the main branch is completed successfully (but before doing the backports) so that in the end people can just check that the diff after rewriting is only this line.

gilles-peskine-arm replied:
My preference would be for you to meld this commit into another one after this PR has been reviewed. Save the original commit in a branch (my personal convention is `cleanup-all.sh-1`) so that Github doesn't delete the commit.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -692,179 +687,173 @@ make CC=gcc CFLAGS='-Werror -Wall -Wextra'
+scripts/config.pl set MBEDTLS_NO_UDBL_DIVISION
+make CC=arm-none-eabi-gcc AR=arm-none-eabi-ar LD=arm-none-eabi-ld CFLAGS='-Werror -O1' lib
+echo "Checking that software 64-bit division is not required"
+! grep __aeabi_uldiv library/*.o
```
Preexisting: if `! grep __aeabi_uldiv library/*.o` fails, it will abort even in keep-going mode. The next line should be `if_build_succeeded ! grep -l __aeabi_uldiv library/*.o`.

mpg replied:
Thanks, will fix.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -692,179 +687,173 @@ make CC=gcc CFLAGS='-Werror -Wall -Wextra'
+
+################################################################
+#### 5. Slow outside Linux-x86_64: MemSan or memcheck
+################################################################
```
Some things are missing here. If you think they're scope creep for this PR, please file an enhancement request on Github.

* The memcheck and memsan builds should run in the full config minus memory management features. We care about uninitialized memory and memory leaks in features that aren't in the default build.
* Valgrind (which we run on buildbot) has spotted memory leaks that Msan didn't spot. So I think we should run both.

mpg replied:
Regarding the first point, I fully agree, so I'll add it to the PR.

Regarding the second, I assume you mean Asan? (Msan only does uninitilised reads, Asan does bounds checks and (on Linux) enables LeakSanitizer.) Do you have an example? I wasn't aware this could happen, I'm surprised to learn that. Regardless, I'd rather leave that to a later stage, for example when we discuss which tests we want in the PR vs nightly/release campaigns.

gilles-peskine-arm replied:
https://github.com/ARMmbed/mbedtls/pull/1545 passed `all.sh` but the [`make memcheck` build found a leak](http://buildbot.polarssl.org/builders/any%20-%20debian-wheezy-i386%20%28cmake%20full%29/builds/994/steps/memcheck/logs/stdio)

mpg commented on tests/scripts/all.sh:
```diff
@@ -551,19 +505,33 @@ cleanup
-if_build_succeeded tests/ssl-opt.sh -f Default
+# test things not in the default config, and Default handshake for parameters choice
+msg "test: ssl-opt.sh (full config)" # < 10s
+if_build_succeeded tests/ssl-opt.sh -f 'Default\|SSLv3\|RC4'
```
Good point. Did a full audit and adjusted the invocation. Turns out only `ECJPAKE` was missing, and `RC4` didn't actually need to be here.

mpg replied:
Good catch, I missed that one. Fixed the name.

mpg commented on tests/scripts/all.sh:
```diff
@@ -174,6 +174,8 @@ fatal_signal () {
 trap 'fatal_signal INT' INT
 trap 'fatal_signal TERM' TERM
+# EXIT covers the case where we abort because of set -e
+trap 'fatal_signal 0' EXIT
```
Oh, I didn't realise that. I didn't have any strong motivation other than a general idea of "let's always clean up", so I'm happy to revert this change if you prefer the old behavioiur.

gilles-peskine-arm replied:
We should have a `--no-ipv6` option. I can add it to #1486.

mazimkhan replied:
> We could use a --no-ipv6 option.

Isn't that ```ssl-opt.sh``` tries to detect if IPv6 is there and skip the tests if not: https://github.com/ARMmbed/mbedtls/blob/development/tests/ssl-opt.sh#L162

mazimkhan replied:
@gilles-peskine-arm thanks for clarifying https://github.com/ARMmbed/mbedtls/pull/1486#issuecomment-378927581

mpg commented on tests/scripts/all.sh:
```diff
@@ -174,6 +174,8 @@ fatal_signal () {
 trap 'fatal_signal INT' INT
 trap 'fatal_signal TERM' TERM
+# EXIT covers the case where we abort because of set -e
+trap 'fatal_signal 0' EXIT
```
Yes, I was lazy and wanted to avoid making `fatal_signal`'s argument optional. Admittedly not the cleanest thing.

mpg commented:
@Patater @gilles-peskine-arm I think I've addressed your feedback, feel free to review again.

@gilles-peskine-arm "Did you test some failing cases?" I'm not sure I understand the question, can you please clarify?

mazimkhan commented on tests/scripts/all.sh:
```diff
@@ -692,179 +687,173 @@ make CC=gcc CFLAGS='-Werror -Wall -Wextra'
-make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS='-Werror' WINDOWS_BUILD=1 tests
-make WINDOWS_BUILD=1 clean
+if [ $RUN_ARMCC -ne 0 ] && [ $YOTTA -ne 0 ]; then
+    # Note - use of yotta is deprecated, and yotta also requires armcc to be on the
```
If user wants to run Yotta tests by specifying ```--yotta``` shouldn't we fail the script when ARMCC is not present. This would silently pass.

mpg replied:
Right. I added a check for that.

mazimkhan commented on tests/scripts/all.sh:
```diff
@@ -692,194 +688,191 @@ make CC=gcc CFLAGS='-Werror -Wall -Wextra'
     msg "build: Release (clang)"
     cleanup
+    scripts/config.pl full
     CC=clang cmake -D CMAKE_BUILD_TYPE:String=Release .
```
Why ```Release``` build is done only when ```MemSan``` is not possible. Are the equivalent alternatives? 
Should we run ```Release``` always? Please document the rationale.

mpg replied:
I think that's a very fair question but I'd rather leave that to a later stage (just like Gilles' point about the fact that we might want valgrind in addition to rather than instead of memsan): as soon as your work on moving the CI scripts to the main repo is done, I expect we'll have a time where we try to rationalize what's covered in the PR vs nightly/release jobs, and I expect that will be a more appropriate time to resolve that.

Currently there is not clear rationale indeed, and one of the goals of this PR is to prepare for some rationalisation work, but not do it right now.

gilles-peskine-arm commented:
By "Did you test some failing cases?", I meant: what testing did you do to check that if a configure/build/test step fails, `all.sh` reported the failure properly? That's always a danger when you change certain parts of the script.

mpg commented:
@gilles-peskine-arm @mazimkhan I believed I've addressed your feedback, please review again.

I suggest leaving the following two questions for later:
- do we need to test with valgrind's memcheck in addition to ASan?
- do we want to make sure we always have a build with cmake in Release mode? (And what about all other cmake modes?)

I'm now running `all.sh -r` locally and will report when it's completed.

mpg commented:
@gilles-peskine-arm I also did some testing by forcing tests to fail (specifically the UDBL one which I think is the only place the PR added a `record_status`), and the failure was reported correctly. However, there was an issue with that test when run without `-k`, which I fixed in the last commit.

I'm now running `all.sh -r` and will then run `all.sh -k -r` and will report on the results.

gilles-peskine-arm commented:
Suggestion: add a PTHREAD build as per #647, and close #647. There's less overhead in merging in a single PR.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -384,7 +384,9 @@ $text"
 else
     record_status () {
-        "$@"
+        if "$@"; then :; else
```
I don't understand this change. It doesn't help to make `record_status ! $command` work. And any change to `record_status` for this purpose needs to also be applied to the keep-going implementation.

Here are several possible solutions:

1. Add `!` detection code to both `record_status` implementations. Replace `"$@"` by

        if [ "$1" = "!" ]; then shift; ! "$@"; else "$@"; fi

2. Instead of `record_status ! …`, use `record_status not …` with

        not () { ! "$@"; }

  (`!` is not a valid function name in many sh implementations.)

3. Instead of `record_status ! …`, use `record_status_not …`, with `record_status_not` implemented for both keep-going and non-keep-going mode.

4. Put the negation in a function:

        check_absence () {
            ! grep "$@"
        }
        if_build_succeeded check_absence __aeabi_uldiv library/*.o

I don't like the duplication in (1) and (3). I have no strong preference between (2) and (4).

mpg replied:
@gilles-peskine-arm This was a misguided attempt as solving the problem, that I should have tested correctly before pushing (I tested it after pushing, but then had no time to fix it before I left the office). I was under the impression, from previous testing, that the "keep going" implementation was working, so I figured `!` only worked in the context of an `if` and decided to aligned to that. Obviously I was very mistaken.

I'll switch to the `not` solution, which I slightly prefer to the `check_absence` one as it is more generic.

gilles-peskine-arm replied:
On second thoughts `not` perpetuates the defect that `grep` could possibly fail for other reasons (very unlikely ones such as a disk error). A more reliable test would be
```
check_absence () {
     ( set +e; grep "$@" ); [ $? -eq 1 ]
 }
```
or
```
check_absence () {
     [ -z "$(grep -l "$@")" ]
 }
```
But I'm fine with `not` too, since I was happy enough with it earlier to commit it and in any case it would be a preexisting issue.

sbutcher-arm commented on README.md:
```diff
@@ -149,7 +149,7 @@ the Mbed TLS source directory, use:
 CMake cache. This can be done with the following command using GNU find:
 
-    find . -iname '*cmake*' -not -name CMakeLists.txt -exec rm -rf {} +
+    find . -name .git -prune -o -iname '*cmake*' -not -name CMakeLists.txt -exec rm -rf {} +
```
This is documentation, not a script. What if the user isn't using git, but is using some other hidden directory that has similar problems? Wouldn't it be better to leave the original command and add a comment to say 'by the way this is destructive, and be sure you're not deleting hidden files that you need, such as in `.git` directories?'

mpg replied:
Good point. I'll leave the command as is (optimising for the most common case) but add a comment that it should be adapted in other cases.

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -149,7 +150,7 @@ cleanup()
     command make clean
 
-    find . -name yotta -prune -o -iname '*cmake*' -not -name CMakeLists.txt -exec rm -rf {} \+
+    find . -name yotta -prune -o -name .git -prune -o -iname '*cmake*' -not -name CMakeLists.txt -exec rm -rf {} \+
```
Just in case other users use subversion or something, as well as having `.git` in the `find`, I think we need a comment.

mpg replied:
I think in the context of `all.sh` this is fine, as we're running git commands two lines below, so clearly we're assuming we're in a git repo already. Also, `all.sh` already has pretty strict environment requirements (assumes two version of OpenSSL and GnuTLS installed, etc.) which means it's only useful on developer or CI machines, so I think this should be acceptable.

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -321,6 +322,12 @@ else
     fi
 fi
 
+if [ $RUN_ARMCC -eq 0 ] && [ $YOTTA -ne 0 ]; then
```
Do we really want to fail on this error? What about `--keep-going`?

mpg replied:
I'd say we want to fail here because:
- this is a user error, not a failing test
- the error is reported early, before any test is run
- this is consistent with how we handle environment checks (which are likely user errors too) just above.

gilles-peskine-arm replied:
Keep-going mode is designed to keep going if there's an expected failure: compilation error, failed test. It isn't designed to keep going in the face of an unexpected failure such as bad command line options or a badly set up environment. So exiting in keep-going mode here is fine.

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -469,101 +467,62 @@ tests/scripts/doxygen.sh
-record_status tests/scripts/test-ref-configs.pl
-
-msg "build: with ASan (rebuild after ref-configs)" # ~ 1 min
-make
```
Why did we have this build and no longer need it?

mpg replied:
We had this build because we chose to run `test-ref-configs.pl` first, but this test does many builds with non-default config, so after running it we needed to re-build with the default config and previous build options before we could go on and run `compat.sh` on the default config with ASan. Now that the `test-ref-configs.pl` has been moved to another section, we only need one build for the whole `make test; ssl-opt.sh; compat.sh` sequence of tests.

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -469,101 +467,62 @@ tests/scripts/doxygen.sh
 msg "test: compat.sh (ASan build)" # ~ 6 min
 if_build_succeeded tests/compat.sh
 
-msg "build: Default + SSLv3 (ASan build)" # ~ 6 min
```
I can see this test along with the other SSLv3 tests have been removed, but it doesn't seem to have been replaced anywhere. Why not? Am I missing something?

mpg replied:
Yes is has been replaced, in the "full config" section (where is it can actually be more thoroughly tested, eg including single-DES suites in SSLv3 mode).

Are you reviewing the full diff or individual commits? I know various people have various preferences about that (and the github interface has a number of issues with per-commit reviews unfortunately) but for this particular PR I'd really recommend reviewing commit by commit, as otherwise the big "reorder" commit makes other changes nearly impossible to follow.

This change is https://github.com/ARMmbed/mbedtls/pull/1243/commits/29a6e3432baf2ec1955c4a405e0454adc715a744 where it's easier to see the added line (and hopefully the commit message should help too).

sbutcher-arm commented:
I'm rather rudely applying the 'not for merging' label until I've completed my review. Apologies for that! Will look at it again as soon as I can. Please don't let the label hold up any other work, and apologies in advance if my questions are a bit dumb. I haven't had enough time to study this properly.

mpg commented:
@sbutcher-arm Thanks for your review! I'm not sure the label was necessary, as I hope nobody was going to merge a PR with an unresolved "changes requested" review anyway. Besides, I previously interpreted this label as meaning that the PR is only there to support discussion, illustrate a concept, etc. while this PR is clearly intended for merging as soon as all reviewers are happy with it... Yet another label we need to clarify ;)

I'll address your feedback soon and am looking forward to the rest of your review!

mpg commented:
@gilles-peskine-arm I'd rather handle #647 as I think answering the (very fair) question you asked there: "Why test the build but not the tests?" could take us too far (if we're testing a threaded build, shouldn't we test `programs/ssl/ssl_server_pthread`? but then, how? shouldn't we use some sort of thread sanitizer? but wait, it has several false positives on our code base (true story)). However, if you agree that just building is a sufficient first step (it's already way better than not even knowing if the code builds), I'll add just that.

@gilles-peskine-arm @sbutcher-arm I believe I've addressed you comments so far (with new commits or replies here), please review again / complete your review.

I'm now locally running two instances (with and without `-k`) and will report on the results later.

gilles-peskine-arm commented:
I'd rather include what's already in #647 and record what's missing in a new issue. But I won't block merging this PR because of that.

mpg commented:
Both `all.sh -r` and `all.sh -r -k` passed locally on c84b39d

mpg commented:
Note: Just edited history in order to simplify the preparation of backports (removed a commit+its revert, squashed commits that fix bugs in previous commits into the original, and squash a commit that only did a cosmetic change into another similar one). 

I've archived the previous history of this PR in this branch: https://github.com/mpg/mbedtls/tree/cleanup-all.sh-0 (tip c84b39d) so that you can easily check that the diff with the new version of this PR is empty.

mpg commented:
Note: I'm quite reluctant to rebase this again unless I have some assurance that it will be reviewed quickly afterwards, as these changes are particularly painful to rebase so I'd rather only do it once.
